### PR TITLE
Exposing transaction FX rate

### PR DESCRIPTION
**Pull request by paramburu**
Following up on discussion: https://github.com/we-promise/sure/discussions/1094

## Overview

This PR adds explicit exchange rate controls to transaction and transfer forms, exposing both **conversion** and **calculated** FX rates when currency differences exist.

## Problem

When users exchange currencies in real life, the final applied rate is often only known at settlement time. By the time users enter the transaction into Maybe, API/reference rates may have changed and no longer match the actual rate used by their bank or exchange provider. This creates:

- Inaccurate account balances
- Incorrect transfer values between multi-currency accounts
- Misleading net worth calculations

This is especially problematic for volatile or spread-heavy currencies like **ARS**, where official/reference rates can differ significantly from the effective rate actually charged due to bank spreads, card rates, taxes, or market conditions.

## Solution

The forms now distinguish between two rate types:

- **Conversion rate**: The real rate actually applied by the bank/exchange provider for this specific transaction (authoritative, saved with the transaction)
- **Calculated rate**: The system/API-derived reference rate used for estimates and defaults (fallback)

Users can now:
- See both rates clearly displayed when currencies differ
- Override the calculated rate with the actual conversion rate used
- Ensure cross-currency transactions and transfers match real-life outcomes

### Backend
- Enhanced `Transaction` and `Transfer::Creator` models to handle conversion rates
- Updated `Balance::SyncCache` to respect user-provided conversion rates
- Extended `Money` class with rate calculation helpers
- Added controller actions for rate previews and calculations

### Frontend
- New Stimulus controllers: `transaction_form_controller.js` and `transfer_form_controller.js`
- Enhanced transaction and transfer forms with rate selection UI
- Real-time rate calculation and validation
- Updated money field component for rate handling

### Localization
- Added i18n keys for rate labels and help text (English and Spanish)

### Testing
- Comprehensive controller tests for rate handling workflows
- Model tests for `Transfer::Creator` rate scenarios
- System tests for end-to-end rate override flows
- `Balance::SyncCache` tests for rate-aware syncing

## Summary by CodeRabbit

* **New Features**
  * Exchange-rate endpoints for transactions and transfers; form UI with Convert/Calculate tabs, live calculations, and conditional display when currencies differ
  * Stimulus-driven interactive behavior for transaction and transfer forms; hidden exchange-rate field shown when needed
  * Allow providing a custom exchange rate when creating transfers and transactions

* **Validation**
  * Exchange rates must be positive

* **Behavior**
  * Money conversion prefers a provided custom rate before fetching or falling back

* **Tests**
  * Expanded coverage for endpoints, UI flows, conversion math, and edge cases

* **Documentation**
  * Added localization strings for the exchange-rate UI and help text

### CODE REVIEW

**Comment by jjmata** [comment]
Very nice addition @paramburu!  Can you review the CodeRabbit/LLM review comments and decide which ones are worth chasing? 🙏

**Comment by jjmata** [comment]
(there's also linter issues it seems ... let's wait for the tests to finish :-) )

**Comment by paramburu** [comment]
> Very nice addition @paramburu! Can you review the CodeRabbit/LLM review comments and decide which ones are worth chasing? 🙏

Already on it. I'll fix the linting issues as well and get back to you.

**Comment by paramburu** [comment]
My bad, forgot to push the fix although I can't run system tests on the devcontainer and still haven't had time to figure out how to sort it. I believe adding [selenium grid](https://hub.docker.com/r/selenium/selenium-grid) with docker compose  within the devcontainer might be a way to sort it out or is there any workaround that I haven't found documented?

File: app/controllers/transfers_controller.rb [hunk 1]
```diff
@@ -8,6 +8,33 @@ def new
...
+
+    if from_account.currency == to_account.currency
+      render json: { rate: nil, same_currency: true }
+    else
+      rate = ExchangeRate.find_or_fetch_rate(
+        from: from_account.currency,
+        to: to_account.currency,
+        date: date
+      )&.rate || 1.0
+
```
[the remark this answers is not in the document]
Comment (paramburu) [reply app/controllers/transfers_controller.rb]: @jjmata I think I’ve hit a bit of a mental block here and would really appreciate some guidance.

The fallback rate of `1` was already in place for transfers. See the existing implementation here: https://github.com/we-promise/sure/blob/ca8f04040f6cb3927c7815a77f42f25997bbcc29/app/models/transfer/creator.rb#L64-L73

That said, I agree that if we’re not careful, a fallback like this could end up persisting incorrect data. My original intention was to normalize everything, but the flow has become more complex since I introduced a dedicated endpoint for fetching exchange rates.

I’ve now added a constraint requiring the exchange rate to be `> 0`, which seems reasonable.

At this point, I’m thinking it might be better to remove the fallback entirely and raise an error if no exchange rate is present—unless it’s being explicitly provided. On the UI side, if no provider is available, we could simply display something like “N/A” instead of enforcing a `> 0` validation.


On the other side we have 

https://github.com/we-promise/sure/blob/ca8f04040f6cb3927c7815a77f42f25997bbcc29/app/models/exchange_rate/provided.rb#L56-L63

which defaults to 1 as well but it's mainly use for reports and such that require "live" data so I won't be changing it. Still, long term we might want to rethink it to make it more obvious that a rate is missing and maybe be able to set it in the settings.

Would that approach make more sense?

[the remark this answers is not in the document]
Comment (jjmata) [reply app/controllers/transfers_controller.rb]: My biggest worry is that transactions flow in outside of the "UI loop" (via import, etc.) - so non-fallback rate (a zero) would make things worse - in theory.

What I mean by "in theory" is that REAL OBVIOUS BREAKAGE is better than "subtle breakage" - no?  But they are both so small (just a few transactions for a few dates, potentially) that they will go unnoticed. Right?

**Comment by paramburu** [comment]
@sokie thanks for the feedback, it's really useful since I am new to the codebase. I already had some changes WIP, I'll work on your feedback and have it rebased and updated by EOW.

**Comment by jjmata** [comment]
Would love to get this into `v0.7.0`! Thank you, Pedro. :pray:

**Comment by paramburu** [comment]
I've taken most of the feedback into account. Let me know if something fell through the cracks.

On the other hand, I've removed the explicit `fallback_rate` path to prevent write-time mis-pricing in multi-currency transaction and transfer creation.

A synthetic fallback (especially `1.0`) can silently persist incorrect ledger amounts. Once written, those amounts propagate into balances, transfer matching, and downstream reporting/analytics.

The conversion path is now explicit and auditable:
1) Use user-provided `custom_rate` when present.
2) Otherwise use provider historical FX for the entry date.
3) If neither exists, fail fast and require manual rate input.

For the sake of simplicity at this time though, we still fall back to an equivalent rate of 1.0 (the original amount) in most read scenarios where a conversion error occurs. We might want to improve that in the future with a note or a visual indicator that something might not be accurate.

**Comment by paramburu** [comment]
Sorry for the test failures. I couldn't run system tests in the devcontainer. I have a fix for it now using selenium docker image. I'll send another PR with that afterwards if it helps.

The only thing I am missing is a failure that seems unrelated, but hasn't failed in the pipeline yet still fails locally running it without my commit, so it might be just a configuration issue or the performance of the container image running on my machine (had to "unparallelize" the tests for them to avoid timeout errors).

```
Failure:
SettingsTest#test_can_update_self_hosting_settings [test/system/settings_test.rb:64]:
expected to find visible css "span[data-clipboard-target=\"iconSuccess\"]" 1 time but there were no matches. Also found "", which matched the selector but not all filters. 
```

**Comment by paramburu** [comment]
Following up on the selenium in devcontainers issue, i've opened PR https://github.com/we-promise/sure/pull/1340

**Comment by paramburu** [comment]
Oh my... sorry again, When splitting the changes between the selenium container and the tests for the this change I moved the wrong method. 🤦

**Comment by paramburu** [comment]
Thanks for the review!
